crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,341 @@
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CoreFetchConfig {
+    /// How many parallel op fetch requests can be made at once. Default: 2.  
+    parallel_request_count: u8,
+    /// Duration in ms to pause when parallel request count is reached. Default: 10.
+    parallel_request_pause: u64,
+    /// Duration in ms to sleep when idle. Default: 1_000.
+    fetch_loop_sleep: u64,
+    /// Duration in ms to keep an unresponsive agent on the cool-down list. Default: 10_000.
+    cool_down_interval: u64,
+}
+
+impl Default for CoreFetchConfig {
+    fn default() -> Self {
+        Self {
+            parallel_request_count: 2,
+            parallel_request_pause: 10,
+            fetch_loop_sleep: 1_000,
+            cool_down_interval: 10_000,
+        }
+    }
+}
+
+impl ModConfig for CoreFetchConfig {}
+
+impl CoreFetchConfig {
+    /// Get parallel request count.
+    pub fn parallel_request_count(&self) -> u8 {
+        self.parallel_request_count
+    }
+    /// Get parallel request pause interval.
+    pub fn parallel_request_pause(&self) -> u64 {
+        self.parallel_request_pause
+    }
+    /// Get fetch loop pause interval.
+    pub fn fetch_loop_pause(&self) -> u64 {
+        self.fetch_loop_sleep
+    }
+    /// Get cool-down interval.
+    pub fn cool_down_interval(&self) -> u64 {
+        self.cool_down_interval
+    }
+}
+
+// TODO: Temporary trait object of a transport module to facilitate unit tests.
+type DynTransport = Arc<Mutex<dyn Transport>>;
+
+#[derive(Debug)]
+struct CoreFetch(Inner);
+
+impl CoreFetch {
+    fn new(config: CoreFetchConfig, transport: DynTransport) -> Self {
+        Self(Inner::spawn_fetch_task(config, transport))
+    }
+}
+
+impl Fetch for CoreFetch {
+    fn add_ops(
+        &mut self,
+        op_list: Vec<OpId>,
+        source: AgentId,
+    ) -> BoxFut<'_, K2Result<()>> {
+        Box::pin(async move {
+            // Add ops to queue.
+            let mut ops = self.0.ops.lock().await;
+            op_list.clone().into_iter().for_each(|op_id| {
+                ops.insert((op_id, source.clone()), ());
+            });
+            drop(ops);
+
+            // Wake up fetch task in case it is sleeping. Time out if the channel buffer
+            // is full.
+            let _ = tokio::time::timeout(
+                Duration::from_millis(5),
+                self.0.fetch_request_tx.send(()),
+            )
+            .await;
+
+            Ok(())
+        })
+    }
+}
+
+#[derive(Clone, Debug)]
+struct Inner {
+    config: CoreFetchConfig,
+    // An `IndexMap` is used to retain order of insertion and have the ability to look up elements
+    // by key efficiently. Ops may be added redundantly to the map with different sources to fetch from, so
+    // the map is keyed by op and agent id together.
+    // The value field could be used to track number of attempts for example.
+    ops: Arc<Mutex<IndexMap<(OpId, AgentId), ()>>>,
```

Comment:
there is an [IndexSet](https://docs.rs/indexmap/latest/indexmap/set/struct.IndexSet.html) if you aren't storing values.

Reply:
I'm struggling to get comfortable with this structure where we have sources for the same op separated. I know we agreed to go simpler on this so I'm fine for now but I'm hoping we'll try to optimize this a bit if we can demonstrate that we're going an unreasonable amount of duplicate fetches later

Reply:
Yes, for sure. For the moment there are assertions for that in the tests even
https://github.com/holochain/kitsune2/pull/35/files#diff-befa5f16b403d32326c69f97ef3ead25c6d7a78f93845c5cda7e6d08d7865a1cR118
and
https://github.com/holochain/kitsune2/pull/35/files#diff-befa5f16b403d32326c69f97ef3ead25c6d7a78f93845c5cda7e6d08d7865a1cR207